couchbase-lite/src/query.rs

```diff
@@ -39,6 +40,37 @@ impl Query<'_> {
             .ok_or_else(|| c4err.into())
     }
 
+    pub(crate) fn new_n1ql<'a, 'b>(db: &'a Database, query_n1ql: &'b str) -> Result<Query<'a>> {
+        let mut c4err = c4error_init();
+        let mut out_error_pos: std::os::raw::c_int = -1;
+        let query = unsafe {
+            c4query_new2(
+                db.inner.0.as_ptr(),
+                kC4N1QLQuery,
+                query_n1ql.as_bytes().as_flslice(),
+                &mut out_error_pos,
+                &mut c4err,
+            )
+        };
+
+        NonNull::new(query)
+            .map(|inner| Query { _db: db, inner })
+            .ok_or_else(|| c4err.into())
+    }
+
+    pub fn set_json_parameters<T>(&self, parameters: &T) -> Result<()> where T: Serialize {
+        let param_string = serde_json::to_string(parameters)?;
+        let param_slice = param_string.as_bytes().as_flslice();
+        unsafe {
+            c4query_setParameters(
+                self.inner.as_ptr(),
+                param_slice
+            );
+        }
+        std::mem::forget(param_string);
```

Comment:
Looks really, really bad. 

We use Rust allocator here, where couchbase-lite-core expect it's own allocator.
So when it try to free memory it will undefined behaviour.
Plus memory leak.

I suppose it should be something like `FLSlice_Copy` call and passing result instead of param_slice

Reply:
Ah I actually kind of expected this type of issues. Thank you. I'll look again.
Thanks.

Reply:
The worse thing that it may work. Because of by default Rust uses system allocator, and C++/new/delete may be also based on system allocator. But then some user of this crate can switch default allocator to jemalloc, or developers of couchbase-lite-core can switch to some other allocator and boom.

I plan someday to write serde fleece serializer/deserializer so it wouldn't be so bad like memory allocation and then another memory allocation. And for document I have to convert it into json5 and then only give to couchbase-lite, which reconvert it into fleece. `c4query_setParameters` looks like support also fleece encoding.

Reply:
Yes that's the reason I named the method set_json_parameters, to leave space for more generic functions.

Reply:
> Yes that's the reason I named the method set_json_parameters, to leave space for more generic functions.

But the imagine optmized fuctnion signature should be exactly like your:

```rust  
pub fn set_json_parameters<T>(&self, parameters: &T) -> Result<()> where T: Serialize {
        let param_enc = serde_fleece::to_fleece_slice(parameters)?;		
        unsafe {
            c4query_setParameters(
                self.inner.as_ptr(),
                params
            );
        }
}
```

the difference would be only in implementation. So why force user to change their code, while we can change our for optimization without change of API?

Reply:
Ah I see (I hadn't seen your comment, sorry).